ranzen/torch/transforms/cutmix.py

```diff
@@ -0,0 +1,181 @@
+        # if p >= 1 then the transform is always applied and we can skip
+        # the above step
+        else:
+            num_selected = batch_size
+            indices = torch.arange(batch_size, device=inputs.device, dtype=torch.long)
+
+        # It's faster to roll the batch by one instead of shuffling it to create image pairs
+        pair_indices = indices.roll(1, 0)
+        lambdas = self.lambda_sampler.sample(sample_shape=torch.Size((num_selected,))).to(
+            inputs.device
+        )
+        lambdas_c = 1.0 - lambdas
+        cutmix_rate = lambdas_c.sqrt()
+        height, width = inputs.shape[-2:]
+        extents_h = torch.round(cutmix_rate * height)
+        extents_w = torch.round(cutmix_rate * width)
+
+        start_inds_h = batched_randint(height - extents_h, generator=self.generator)
+        start_inds_w = batched_randint(width - extents_w, generator=self.generator)
+
+        end_inds_h = start_inds_h + extents_h
+        end_inds_w = start_inds_w + extents_w
+
+        # Convert the bounding box coordinates into masks.
+        range_h = torch.arange(height, device=inputs.device).unsqueeze(0).expand(num_selected, -1)
+        masks_h = (end_inds_h.unsqueeze(-1) >= range_h) & (range_h >= start_inds_h.unsqueeze(-1))
+        range_w = torch.arange(width, device=inputs.device).unsqueeze(0).expand(num_selected, -1)
+        masks_w = (end_inds_w.unsqueeze(-1) >= range_w) & (range_w >= start_inds_w.unsqueeze(-1))
+        masks = (
+            (masks_h.unsqueeze(1) * masks_w.unsqueeze(-1))
+            .unsqueeze(1)
+            .expand(-1, inputs.size(1), -1, -1)
+        )
+
+        if not self.inplace:
+            inputs = inputs.clone()
+        inputs[indices] = ~masks * inputs[indices] + masks * inputs[pair_indices]
+        # No targets were recevied so we're done.
+        if targets is None:
+            return inputs
+
+        # Targets are label-encoded and need to be one-hot encoded prior to mixup.
+        if torch.atleast_1d(targets.squeeze()).ndim == 1:
+            if self.num_classes is None:
+                raise RuntimeError(
+                    f"{self.__class__.__name__} can only be applied to label-encoded targets if "
+                    "'num_classes' is specified."
+                )
+            targets = cast(Tensor, F.one_hot(targets, num_classes=self.num_classes))
+        elif not self.inplace:
+            targets = targets.clone()
+        # Targets need to be floats to be mixed up.
+        targets = targets.float()
+        target_lambdas = 1.0 - (end_inds_w - start_inds_w) * (end_inds_h - start_inds_h) / (
+            width * height
+        )
+        target_lambdas.unsqueeze_(-1)
+        targets[indices] *= target_lambdas
+        targets[indices] += targets[pair_indices] * (1.0 - target_lambdas)
+
+        return InputsTargetsPair(inputs=inputs, targets=targets)
+
+    @overload
+    def __call__(self, inputs: Tensor, *, targets: Tensor = ...) -> InputsTargetsPair:
```

Comment:
```suggestion
    def __call__(self, inputs: Tensor, *, targets: Tensor) -> InputsTargetsPair:
```

Reply:
Why does this apply to the ``Tensor`` overload but not the`` None`` one?

Reply:
Only one overload should have a default value. Because if both have a default value and the call site elides that argument, then how could pyright tell which overload to use?

Reply:
Right, I was able to divine your reasoning but it seems to make no difference in practice -- pyright correctly infers the return type whether all of the overloads are give placeholder defaults or just the one.

Reply:
Oh never mind, your rule does apply in the absence of a second argument, obviously

Reply:
I think this is another place where pyright is more forgiving than mypy. I think mypy would always complain about this.

Reply:
Hmm pyright may well do if strict-checking is enabled (which it currently isn't for ranzen)